etna/transforms/timestamp/fourier.py

```diff
@@ -221,22 +294,41 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
         result:
             transformed dataframe
         """
+        if self._freq is _DEFAULT_FREQ:
+            raise ValueError("The transform isn't fitted!")
+
         if self.in_column is None:
             if pd.api.types.is_integer_dtype(df.index.dtype):
                 timestamp = df.index.to_series()
             else:
-                timestamp = self._convert_sequential_timestamp_datetime_to_int(timestamp=df.index)
+                timestamp = self._convert_regular_timestamp_datetime_to_numeric(
+                    timestamp=df.index, reference_timestamp=self._reference_timestamp, freq=self._freq
+                )
             features = self._compute_features(timestamp=timestamp)
             features.index = df.index
             result = self._construct_answer_for_index(df=df, features=features)
         else:
-            flat_timestamp_df = TSDataset.to_flatten(df=df, features=[self.in_column])
-
+            # here we are assuming that every segment has the same timestamp dtype
             timestamp_dtype = df.dtypes.iloc[0]
             if pd.api.types.is_numeric_dtype(timestamp_dtype):
+                flat_timestamp_df = TSDataset.to_flatten(df=df)
                 timestamp = flat_timestamp_df[self.in_column]
             else:
-                raise ValueError("Only numeric data is supported if in_column is set!")
+                self._validate_external_timestamp(df=df)
+                segments = df.columns.get_level_values("segment").unique()
+                int_timestamp_values = []
+                for segment in segments:
```

Comment:
Could we make it without loops?

Reply:
I think it is possible, but I haven't figured an easy way to do this, because we need to run `pd.date_range` in parallel.

More efficient implementation requires more work, and I'm not sure is it worth it or not.